compiler/compiler.go

```diff
@@ -277,19 +255,98 @@ func WritePkgCode(pkg *Archive, dceSelection map[*Decl]struct{}, gls goLinknameS
 	return nil
 }
 
+type serializableArchive struct {
+	ImportPath   string
+	Name         string
+	Imports      []string
+	ExportData   []byte
+	Declarations []*Decl
+	IncJSCode    []byte
+	FileSet      []byte
+	Minified     bool
+	GoLinknames  []GoLinkname
+	BuildTime    time.Time
+}
+
 // ReadArchive reads serialized compiled archive of the importPath package.
-func ReadArchive(path string, r io.Reader) (*Archive, error) {
+//
+// The given srcModTime is used to determine if the archive is out-of-date.
+// If the archive is out-of-date, the returned archive is nil.
+// If there was not an error, the returned time is when the archive was built.
+//
+// The imports map is used to resolve package dependencies and may modify the
+// map to include the package from the read archive. See [gcexportdata.Read].
+func ReadArchive(importPath string, r io.Reader, srcModTime time.Time, imports map[string]*types.Package) (*Archive, time.Time, error) {
+	var sa serializableArchive
+	if err := gob.NewDecoder(r).Decode(&sa); err != nil {
+		return nil, time.Time{}, err
+	}
+
+	if srcModTime.After(sa.BuildTime) {
+		// Archive is out-of-date.
+		return nil, sa.BuildTime, nil
+	}
```

Comment:
From the separation of concerns perspective, I would expect ReadArchive() to always return the archive content, regardless of its age, and then its validity for a particular purpose to be decided separately.

Reply:
The reason was to try to keep the same functionality. Before the change we didn't deserialize the package and register the type until after checking that the Archive was up-to-date. Since [`gcexportdata.Read`](https://pkg.go.dev/golang.org/x/tools/go/gcexportdata#Read) says "Read may inspect and add to the imports map" I was worried what would happen if we deserialized a package that was out-of-date, then tried to overwrite it later with an up-to-date version.

I also wasn't sure of what changes, if any, it'd make to the import packages and types. If the only changes are to the import map, then we could make a copy of that map before the Read changes it, then only if the package is not old, merge the imports map while registering the type. Since I wasn't sure, I didn't want to risk it.

Anyway, I'm cool with whatever, including undoing those changes and leaving the package as serialized bytes. Since the package was already serialized, everywhere that needed the package just looks it up in the imports map using the path (as far as I could tell), so it's not like it is required to have that deserialized right now.

Reply:
Ah, right, I did not take the possible side-effects into account. Yeah, in this case such a behavior is justified. Maybe note it down somewhere in a comment for future archaeologists? :)